### PR TITLE
BGD-3078 - Error message improvement

### PR DESCRIPTION
**Pull request by arnarpall**
Will now report and log error codes properly when failues occur when submitting an app for example.

We will also now expose the `requestID` so that we can better assist when an error occurs.

Changed the retry policy for connections so that we only retry on `ConnectionErrors` and `Timeouts` 
In most cases it's not really helpful to retry errors that we get from our platform since they are not recoverable.
Mostly these are validation errors, none existing clusters e.t.c

### CODE REVIEW

File: ocean_spark/hooks.py [hunk 1]
```diff
@@ -92,38 +96,38 @@ def _do_api_call(
...
-                    # In this case, the user probably made a mistake.
-                    # Don't retry.
-                    raise AirflowException(
-                        "Response: {0}, Status Code: {1}".format(
-                            e.response.content, e.response.status_code
-                        )
-                    )
-
-                self._log_request_error(attempt_num, e)
+                api_response: ApiResponse = json.loads(e.response.content)
```
[the remark this answers is not in the document]
Comment (arnarpall) [reply ocean_spark/hooks.py:109]: Yeah I agree.

[the remark this answers is not in the document]
Comment (arnarpall) [reply ocean_spark/hooks.py]: @antoinewdg-netapp  I've updated handling.

File: ocean_spark/hooks.py [hunk 2]
```diff
@@ -92,37 +95,49 @@ def _do_api_call(
...
+            errors = api_response["response"]["errors"]
+            status_code = api_response["response"]["status"]["code"]
+            self.log.error(
+                "Request %s, to Ocean Spark API failed with errors: %s",
+                request_id,
+                errors,
+            )
+            raise AirflowException(
+                f"RequestID: {request_id}, errors: {errors}, Status Code: {status_code}"
+            )
```
Comment (arnarpall) [inline ocean_spark/hooks.py:133]: I'm not a huge fan of the `log-rethrow` pattern being used here.

I think I might just remove the log line and relyon the calling code to log the exception properly. As seem  to be the case in airflow anyway.

Comment (arnarpall) [reply ocean_spark/hooks.py]: Tried this approach and it does not work.

File: ocean_spark/hooks.py [hunk 3]
```diff
@@ -92,37 +95,38 @@ def _do_api_call(
...
-    ) -> None:
-        self.log.error(
-            "Attempt %s API Request to Data Mechanics failed with reason: %s",
-            attempt_num,
-            error,
-        )
+    @staticmethod
+    def _construct_error_message(ex: requests_exceptions.RequestException) -> str:
+        try:
+            api_response: ApiResponse = json.loads(ex.response.content)
```
Comment (antoinewdg-netapp) [inline ocean_spark/hooks.py:123]: Are you sure the response is always going to contain valid JSON?

For example, what if the HTTP error code comes from a load balancer on top of the optimizer whose connection to the optimizer timed out?

File: ocean_spark/hooks.py [hunk 4]
```diff
@@ -92,37 +95,38 @@ def _do_api_call(
...
+                self.log.error(msg)
+                raise AirflowException(msg)
 
             if attempt_num == self.retry_limit:
                 raise AirflowException(
-                    (
-                        "API requests to Data Mechanics failed {} times. "
-                        + "Giving up."
-                    ).format(self.retry_limit)
+                    f"Request to Ocean Spark API failed failed {self.retry_limit} times. Giving up."
```
Comment (thorsteinnth) [inline ocean_spark/hooks.py:114]: failed failed

File: ocean_spark/hooks.py [hunk 5]
```diff
@@ -92,37 +95,38 @@ def _do_api_call(
...
+    @staticmethod
+    def _construct_error_message(ex: requests_exceptions.RequestException) -> str:
+        try:
+            api_response: ApiResponse = json.loads(ex.response.content)
+            request_id = api_response["request"]["id"]
+            errors = api_response["response"]["errors"]
+            status_code = api_response["response"]["status"]["code"]
+            return f"Request '{request_id}' to the Ocean Spark API failed with status '{status_code}' and errors: {errors}"
+        except ValueError:
+            return f"Request to the Ocean for Spark API failed with status '{ex.response.status_code}' and unknown errors {ex.response.content}"
```
Comment (thorsteinnth) [inline ocean_spark/hooks.py:129]:
```suggestion
            return f"Request to the Ocean Spark API failed with status '{ex.response.status_code}' and error {ex.response.content}"
```

File: ocean_spark/hooks.py (hunk 4 above)
Comment (thorsteinnth) [inline ocean_spark/hooks.py:114]:
```suggestion
                    f"Request to Ocean Spark API failed {self.retry_limit} times. Giving up."
```

**Review by thorsteinnth — APPROVED** [review]

**Pull request merged**